tools/block-generator/generator/generate.go

```diff
@@ -369,90 +299,59 @@ func (g *generator) WriteBlock(output io.Writer, round uint64) error {
 		return nil
 	}
 	// round == nextRound case
+
 	err := g.startRound()
 	if err != nil {
 		return err
 	}
-	numTxnForBlock := g.txnForRound(g.round)
+	if g.round == 0 {
+		fmt.Printf("starting txnCounter: %d\n", g.txnCounter)
+	}
+	minTxnsForBlock := g.minTxnsForBlock(g.round)
 
-	var intra uint64 = 0
 	var cert rpcs.EncodedBlockCert
 	if g.round == 0 {
 		// we'll write genesis block / offset round for non-empty database
 		cert.Block, _, _ = g.ledger.BlockCert(basics.Round(round - g.roundOffset))
 	} else {
-		// generate a block
-		cert.Block.BlockHeader = bookkeeping.BlockHeader{
-			Round:          basics.Round(g.round),
-			Branch:         bookkeeping.BlockHash{},
-			Seed:           committee.Seed{},
-			TxnCommitments: bookkeeping.TxnCommitments{NativeSha512_256Commitment: crypto.Digest{}},
-			TimeStamp:      g.timestamp,
-			GenesisID:      g.genesisID,
-			GenesisHash:    g.genesisHash,
-			RewardsState: bookkeeping.RewardsState{
-				FeeSink:                   g.feeSink,
-				RewardsPool:               g.rewardsPool,
-				RewardsLevel:              0,
-				RewardsRate:               0,
-				RewardsResidue:            0,
-				RewardsRecalculationRound: 0,
-			},
-			UpgradeState: bookkeeping.UpgradeState{
-				CurrentProtocol: g.protocol,
-			},
-			UpgradeVote:        bookkeeping.UpgradeVote{},
-			StateProofTracking: nil,
-		}
+		g.setBlockHeader(&cert)
 
-		// Generate the transactions
-		transactions := []txn.SignedTxnInBlock{}
-		for intra < numTxnForBlock {
-			var signedTxns []txn.SignedTxn
-			var ads []txn.ApplyData
-			var err error
-			signedTxns, ads, intra, err = g.generateSignedTxns(g.round, intra)
+		intra := uint64(0)
+		txGroupsAD := [][]txn.SignedTxnWithAD{}
+		for intra < minTxnsForBlock {
+			txGroupAD, numTxns, err := g.generateTxGroup(g.round, intra)
```

Comment:
Should the txgroup stuff get pushed into `g.evaluateBlock`?

Reply:
Perhaps lines 319-332 could become:
```go
intra, txGroupsAD := g.generateTxGroups()
```

Or perhaps lines 319-334 could become:
```go
vBlock, ledgerTxnCount, err := g.generateBlock()
```

Any other ideas?

Reply:
generating the tx groups is part of generating the block, so I was thinking you could push 319-340 into `g.evaluateBlock`, and rename `evaluateBlock` to `generateBlock`